lambda/helper/python/comprehendHelper.py

```diff
@@ -0,0 +1,454 @@
+        MAX_COMPREHEND_UTF8_PAGE_SIZE = 5000
+        
+        # the textract results json file in S3 has a peculiar structure of
+        # top-level list of block lists, each appear to have a size limit of 1000.
+        # A new one is creatd and appended when the previous blocklist reaches that limit.
+        # Not sure why it is so. A blocklist may contain blocks from different document pages.
+        for blocklist in textract:
+        
+            for block in blocklist['Blocks']:
+        
+                # PAGE block type have no text, so skip those
+                if block['BlockType'] != 'PAGE':
+                    
+                    # page numbers start at 1 in Textract for png, however
+                    # for png there are no page reference since it is a single
+                    # page, in this case all blocks have page 1
+                    if numOfPages == 1:
+                        pageNumber = 1
+                    else:
+                        pageNumber = block['Page']
+                    
+                    # skip pags above limit
+                    if pageNumber > numOfPages:
+                        continue
+                    
+                    # but our storage of page results list start at index 0
+                    pageResultIndex = pageNumber - 1
+                
+                    # some block may not contain text
+                    if  'Text' in block:
+                        
+                        # calculate the size of this page if we add this text element + a space separator.
+                        # Comprehend has a UTF8 size limit, so we dismiss excessive elements once size is
+                        # reached.
+                        projectedSize = len(rawPages[pageResultIndex]) + len(block['Text']) + 1
+                        
+                        # add if page size allows
+                        if MAX_COMPREHEND_UTF8_PAGE_SIZE > projectedSize :
+                            # add a separator from previous text block
+                            rawPages[pageResultIndex] += " "
+                            # text block
+                            rawPages[pageResultIndex] += block['Text']
+
+        return numOfPages
+
+    #
+    # thread execution calling Comprehend synchronously by batch of up to
+    # 25 pages
+    #
+    def batchComprehendDetectEntitiesSync(self,
+                                          rawPages,
+                                          pagesToProcess,
+                                          pageStartIndex,
+                                          comprehendEntities):
+        
+        try:
+            client = boto3.client('comprehend')
+        
+            textPages = []
+            endIndex = pageStartIndex + pagesToProcess
+        
+            for i in range(pageStartIndex, endIndex):
+                textPages.append(rawPages[i])
+
+            # service limit is around 10tps, sdk implements 3 retries with backoff
```

Comment:
Here is an interesting [issue](https://github.com/boto/botocore/issues/1946) I found. Seems boto doesnt take care of all retryable exceptions.

Reply:
whoah...   we could handle that but it will get fixed eventually and I'm reluctant to put code that customers will see to that effect